### PR TITLE
added JWT support

### PR DESCRIPTION
**Pull request by Nazar090**

### CODE REVIEW

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
build failed, please fix it first

File: src/main/java/com/example/onlinebookstore/config/OpenApiConfig.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    @Bean
+    public OpenAPI customOpenApi() {
+        return new OpenAPI()
+            .components(new Components().addSecuritySchemes("BearerAuth",
+                new SecurityScheme()
+                  .type(SecurityScheme.Type.HTTP)
+                  .scheme("bearer")
+                  .bearerFormat("JWT")))
+            .addSecurityItem(new SecurityRequirement().addList("BearerAuth"));
```
Comment (MentorMateAcademy) [inline src/main/java/com/example/onlinebookstore/config/OpenApiConfig.java]: Save strings to constants

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                    userDetails, null, userDetails.getAuthorities()
+            );
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
```
Comment (MentorMateAcademy) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]: Constants

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/onlinebookstore/controller/AuthController.java [hunk 3]
```diff
@@ -19,11 +22,18 @@
...
     @PostMapping("/register")
     @Operation(summary = "User registration", description = "User registration in the store")
     public UserResponseDto registerUser(@RequestBody @Valid UserRegistrationRequestDto requestDto)
             throws RegistrationException {
         return userService.register(requestDto);
     }
+
+    @PostMapping("/login")
+    @Operation(summary = "User login", description = "User login in the store")
+    public UserLoginResponseDto login(@RequestBody UserLoginRequestDto request) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/controller/AuthController.java]:
```suggestion
    public UserLoginResponseDto login(@RequestBody @Valid UserLoginRequestDto request) {
```

File: src/main/java/com/example/onlinebookstore/config/SecurityConfig.java [hunk 4]
```diff
@@ -29,7 +34,7 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
             .csrf(AbstractHttpConfigurer::disable)
             .authorizeHttpRequests(
                     auth -> auth
-                            .requestMatchers("/auth/register",
+                            .requestMatchers("/auth/login", "/auth/register",
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/config/SecurityConfig.java]:
```suggestion
                            .requestMatchers("/auth/**",
```

File: src/main/java/com/example/onlinebookstore/dto/user/UserLoginRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package com.example.onlinebookstore.dto.user;
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotEmpty;
+import jakarta.validation.constraints.Size;
+
+public record UserLoginRequestDto(
+        @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/user/UserLoginRequestDto.java]:
```suggestion
        @NotBlank
```

File: src/main/java/com/example/onlinebookstore/dto/user/UserLoginRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+        @Size(min = 8, max = 35)
+        @Email
+        String email,
+        @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/user/UserLoginRequestDto.java]:
```suggestion
        @NotBlank
```

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+@Component
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private static final String tokenStartsWith = "Bearer ";
+    private static final String authCredentials = "Authorization";
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+            Authentication authentication = new UsernamePasswordAuthenticationToken(
+                    userDetails, null, userDetails.getAuthorities()
+            );
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader(authCredentials);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
        String bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
```

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+@Component
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private static final String tokenStartsWith = "Bearer ";
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
    private static final String TOKEN_HEADER = "Bearer ";
```

constants should contain only upper case letters

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            );
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader(authCredentials);
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith(tokenStartsWith)) {
+            return bearerToken.substring(7);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
            return bearerToken.substring(TOKEN_HEADER.length());
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            Authentication authentication = new UsernamePasswordAuthenticationToken(
+                    userDetails, null, userDetails.getAuthorities()
+            );
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader(AUTHORIZATION);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
        String bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
```

use Spring constants from import org.springframework.http.HttpHeaders;

File: src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+@Component
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends OncePerRequestFilter {
+    private static final String TOKEN_HEADER = "Bearer ";
+    private static final String AUTHORIZATION = "Authorization";
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/security/JwtAuthenticationFilter.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
just one comment

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**